File: birthday.py
from bitcoin.rpc import RawProxy
import datetime

proxy = RawProxy()

def binary_search(max_block_height, target_time):
    # left, right = 0, len(arr) - 1
    # print(max_block_height)
    left_block_height, right_block_height = 0, max_block_height

    while left_block_height <= right_block_height:
        mid_block_height = (left_block_height + right_block_height) // 2

        # print(mid_block_height)
        mid_block = proxy.getblock(proxy.getblockhash(mid_block_height))
        left_block = proxy.getblock(proxy.getblockhash(left_block_height))
        right_block = proxy.getblock(proxy.getblockhash(right_block_height))

        mid_block_time = mid_block["time"]
        left_block_time = left_block["time"]
        right_block_time = right_block["time"]

        # print (mid_block_time)
        # print (left_block_time)
        # print (right_block_time)

        # print()

        # Check if target is present at mid
        if mid_block_time == target_time:
            return mid_block_height
        # If target is greater, ignore left half
        elif mid_block_time < target_time:
            left_block_height = mid_block_height + 1
        # If target is smaller, ignore right half
        else:
            right_block_height = mid_block_height - 1

    # If the target is not found in the array
    return mid_block_height

# get max block height
block = proxy.getblock(proxy.getbestblockhash())
max_block_height = block["height"]

# get target time
year = int(input('Enter year:\n'))
month = int(input('Enter month:\n'))
day = int(input('Enter day:\n'))
hour = int(input('Enter hour:\n'))
minute = int(input('Enter minute:\n'))
sec = int(input('Enter second:\n'))

# given_datetime = datetime.datetime(2020, 2, 18, 15, 38, 7)
given_datetime = datetime.datetime(year, month, day, hour, minute, sec)

# 2022, 6, 3, 15, 7, 0

target_time = int(given_datetime.timestamp())
# print(type(target_time))
# print(target_time)
# target_time = 1582056227
# target_time = 1654283220

# find result
print("Finding blockheight...")
result = binary_search(max_block_height, target_time)
print(f'The blockheight at this date and time was: {result}')

# if result != -1:
#     print(f"Element {target} is present at index {result}")
# else:
#     print(f"Element {target} is not present in the array")

File: birthday.go
// Copyright (c) 2014-2017 The btcsuite developers
// Use of this source code is governed by an ISC
// license that can be found in the LICENSE file.

package main

import (
	"log"
	"fmt"
	"time"
    "os"
	"github.com/btcsuite/btcd/rpcclient"
    "github.com/joho/godotenv"
)

func binarySearch(client *rpcclient.Client, blockCount int64, targetTime int64) int64 {
	var leftBlockHeight, rightBlockHeight int64 = 0, blockCount

	for leftBlockHeight <= rightBlockHeight {

		// fmt.Println(rightBlockHeight)
		// fmt.Println(blockCount)
		// fmt.Println(targetTime)

		midBlockHeight := (leftBlockHeight + rightBlockHeight) / 2

		midBlockHash, err := client.GetBlockHash(midBlockHeight)
			if err != nil {
				log.Fatal(err)
			}

		midBlock, err := client.GetBlockVerbose(midBlockHash)
			if err != nil {
				log.Fatal(err)
			}
		// fmt.Println(midBlock.Time)

		// leftBlockHash, err := client.GetBlockHash(leftBlockHeight)
		// 	if err != nil {
		// 		log.Fatal(err)
		// 	}

		// leftBlock, err := client.GetBlockVerbose(leftBlockHash)
		// 	if err != nil {
		// 		log.Fatal(err)
		// 	}

		// rightBlockHash, err := client.GetBlockHash(rightBlockHeight)
		// 	if err != nil {
		// 		log.Fatal(err)
		// 	}

		// rightBlock, err := client.GetBlockVerbose(rightBlockHash)
		// 	if err != nil {
		// 		log.Fatal(err)
		// 	}

		midBlockTime := midBlock.Time
		// leftBlockTime := leftBlock.Time
		// rightBlockTime := rightBlock.Time

		if midBlockTime == targetTime {
			return midBlockHeight
		} else if midBlockTime < targetTime {
			leftBlockHeight = midBlockHeight + 1
		} else {
			rightBlockHeight = midBlockHeight - 1
		}
	}
	return(leftBlockHeight)

}

func main() {
    // Load the environment variables from the .env file
    err := godotenv.Load(".env")
    if err != nil {
        log.Fatalf("Error loading .env file")
    }
    username := os.Getenv("BTCUSER")
    password := os.Getenv("PASSWORD")
	// Connect to local bitcoin core RPC server using HTTP POST mode.
	connCfg := &rpcclient.ConnConfig {
		Host:         "localhost:8332",
		User:         username,
		Pass:         password,
		HTTPPostMode: true, // Bitcoin core only supports HTTP POST mode
		DisableTLS:   true, // Bitcoin core does not provide TLS by default
	}
	// Notice the notification parameter is nil since notifications are
	// not supported in HTTP POST mode.
	client, err := rpcclient.New(connCfg, nil)
	if err != nil {
		log.Fatal(err)
	}
	defer client.Shutdown()

	// Get the current block count.
	blockCount, err := client.GetBlockCount()
	if err != nil {
		log.Fatal(err)
	}
	// log.Printf("Block count: %d", blockCount)

	var year, month, day, hour, minute, sec int
	fmt.Println("Note: assuming eastern time")
	fmt.Println("Enter year:")
	fmt.Scan(&year)
	fmt.Println("Enter month:")
	fmt.Scan(&month)
	fmt.Println("Enter day:")
	fmt.Scan(&day)
	fmt.Println("Enter hour:")
	fmt.Scan(&hour)
	fmt.Println("Enter minute:")
	fmt.Scan(&minute)
	fmt.Println("Enter second:")
	fmt.Scan(&sec)

	location, err := time.LoadLocation("America/New_York")
	if err != nil {
		// Handle error, e.g., log it or return an error
		fmt.Println("Error loading location:", err)
		// return or handle the error as appropriate
	}

	givenDateTime := time.Date(year, time.Month(month), day, hour, minute, sec, 0, location)
	// fmt.Printf("given date time: %d\n", givenDateTime)
	// fmt.Printf("given date time: %d\n", givenDateTime.UTC())
	targetTime := givenDateTime.Unix()
	// fmt.Printf("target time: %d\n", targetTime)


	fmt.Println("Finding block height...")
	result := binarySearch(client, blockCount, targetTime)
	// fmt.Println(result)
	fmt.Printf("The block height at this date and time was: %d\n", result)
}
